### ISSUE TITLE
DataTable: Custom State Restore causes data to be loaded multiple times when using lazy (remote data)

### ISSUE THREAD

**Issue by nicholas-brooks** [primefaces/primereact#6138]
### Describe the bug

Using Custom State management via stateStorage="custom" and customSaveState and customRestoreState causes data to be fetched multiple times when restoring the state when using "lazy" data (fetching data from an external source).

This appears to be caused by the restore state code calling onPage, onSort, onFilter, etc. individually as it is restoring the internal state.  As our code (and the provided example) refetches data when onPage, onSort, onFilter, etc. is called, this causes data to be fetched multiple times.

What is worse is that each time the data is fetched,  it is fetched with progressively more state information applied.

e.g.

onPage is called - we only know what page to fetch data for.
onSort is called - we now know the page as well as the sorting to fetch data for.
onFilter is called - we now know the page, sorting and filtering criteria to fetch data for.

This gif shows the issue we're seeing - check out the dev console as well (gif taken from the reproducer example):

![2024-03-13 16 11 48](https://github.com/primefaces/primereact/assets/1787619/d4973712-e4ea-4002-bd1b-a1c53bbaf0d4)

### Reproducer

https://stackblitz.com/edit/1jn8yc

### PrimeReact version

10.5.1

### React version

18.x

### Language

ALL

### Build / Runtime

Vite

### Steps to reproduce the behavior

Using the reproducer example:
1. Navigate to index
2. Navigate to Data Table
3. Filter the data table.
4. Navigate back to index.
5. Navigate back to Data Table. This will cause Data Table to restore the state and thus demonstrate the issue.

### Expected behavior

In the reproducer example, lazyLoading should only be executed after all of the state of the DataTable is restored.  How that is achieved, I don't know.

**Comment by emreakdas** [comment]
@nicholas-brooks did you solve the problem?

**Comment by nicholas-brooks** [comment]
@emreakdas No.  Also the problem still exists in latest 10.6.2 (just checked).  I cannot resolve the problem outside of modifying DataTable source as the issue is with how the _restoreState method works: https://github.com/primefaces/primereact/blob/308e350423d177e5c6ca2964c888141a65d8a5a3/components/lib/datatable/DataTable.js#L288

Issue is that _restoreState is causing re-renders each time it restores an individual piece of the saved state (e.g. sorting).  Ideally it should restore all of the state that it has saved previously, then and only then tell the outside world that state has changed.  How?  No idea.  Haven't had time to delve into it much more than that.

**Comment by emreakdas** [comment]
@nicholas-brooks thanks for your comment, I was having the same problem in resizing the table etc. The problem was overcome with useCallback, but the same problem exists in sorting or filter. I will have to find other workarounds

**Comment by emreakdas** [comment]
@mertsincan cc
